sample-tracking-status-overview-app/src/main/groovy/life/qbic/portal/sampletracking/datasources/samples/SamplesDbConnector.groovy

```diff
@@ -35,6 +35,41 @@ class SamplesDbConnector implements CountSamplesDataSource {
         this.connectionProvider = connectionProvider
     }
 
+    /**
+     * {@inheritDoc}
+     * @since 1.0.0
+     * We need to fetch all samples initially, since only the latest status counts. This only remains true, while
+     * Status.DATA_AVAILABLE is the last status a sample can reach.
+     */
+    @Override
+    List<String> fetchSampleCodesWithData(String projectCode) throws DataSourceException {
+        String queryTemplate = Query.fetchLatestSampleEntries()
+        Connection connection = connectionProvider.connect()
+        List<Status> statuses = new ArrayList<>()
+        String sqlRegex = "$projectCode%"
+        connection.withCloseable {
+            PreparedStatement preparedStatement = it.prepareStatement(queryTemplate)
+            preparedStatement.setString(1, sqlRegex)
+            ResultSet resultSet = preparedStatement.executeQuery()
+            while (resultSet.next()) {
+                String sampleCode = resultSet.getString("sample_id")
+                String sampleStatusString = resultSet.getString("sample_status")
+                String arrivalTime = resultSet.getString("arrival_time")
+                Status sampleStatus
+                try {
+                    sampleStatus = Status.valueOf(sampleStatusString)
+                } catch(IllegalArgumentException statusNotFound) {
+                    // The status in the database is invalid. This should never be the case!
+                    log.error("Could not parse status $sampleStatusString for $sampleCode at $arrivalTime", statusNotFound)
+                    throw new DataSourceException("Retrieval of sample statuses failed for sample $sampleCode")
+                }
+                if(Status.DATA_AVAILABLE.equals(sampleStatus)) {
```

Comment:
Can we be sure that the available data status was not changed at a later time? otherwise I would find it hard to incorporate that into a query that also takes the latest sample status.

Reply:
As the status itself (DATA_AVAILABLE) is not expected to change in its meaning I would incorporate it into the query. Especially, since you are in the app layer and expected to deliver available statuses only. 😄 Otherwise you would have to do it like before and retrieve all data and filter it in the domain layer